src/loader/extension.c

```diff
@@ -0,0 +1,198 @@
+#include <postgres.h>
+#include <access/xact.h>
+#include <commands/extension.h>
+#include <catalog/namespace.h>
+#include <utils/lsyscache.h>
+#include <miscadmin.h>
+#include <parser/analyze.h>
+#include <commands/extension.h>
+#include <access/relscan.h>
+#include <catalog/pg_extension.h>
+#include <utils/fmgroids.h>
+#include <utils/builtins.h>
+#include <utils/rel.h>
+#include <catalog/indexing.h>
+
+#include "extension.h"
+
+#define EXTENSION_PROXY_TABLE "cache_inval_extension"
+#define CACHE_SCHEMA_NAME "_timescaledb_cache"
+#define MAX_SO_NAME_LEN NAMEDATALEN+NAMEDATALEN+1+1		/* extname+"-"+version */
+
+static bool loaded = false;
+
+/* This is the extension's post_parse_analyze_hook */
+static post_parse_analyze_hook_type extension_post_parse_analyze_hook = NULL;
+
+
+static bool inline
+proxy_table_exists()
+{
+	Oid			nsid = get_namespace_oid(CACHE_SCHEMA_NAME, true);
+	Oid			proxy_table = get_relname_relid(EXTENSION_PROXY_TABLE, nsid);
+
+	return OidIsValid(proxy_table);
+}
+
+static bool inline
+extension_exists()
+{
+	return OidIsValid(get_extension_oid(EXTENSION_NAME, true));
+}
+
+static bool inline
+extension_is_transitioning()
+{
+	/*
+	 * Determine whether the extension is being created or upgraded (as a
+	 * misnomer creating_extension is true during upgrades)
+	 */
+	if (creating_extension)
+	{
+		char	   *current_extension_name = get_extension_name(CurrentExtensionObject);
+
+		if (NULL == current_extension_name)
+		{
+			elog(ERROR, "Unknown current extension while creating");
+		}
+
+		if (strcmp(EXTENSION_NAME, current_extension_name) == 0)
+		{
+			return true;
+		}
+	}
+	return false;
+}
+
+static char *
+extension_version(void)
+{
+	Datum		result;
+	Relation	rel;
+	SysScanDesc scandesc;
+	HeapTuple	tuple;
+	ScanKeyData entry[1];
+	bool		is_null = true;
+	static char *sql_version = NULL;
+
+	rel = heap_open(ExtensionRelationId, AccessShareLock);
+
+	ScanKeyInit(&entry[0],
+				Anum_pg_extension_extname,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(EXTENSION_NAME));
+
+	scandesc = systable_beginscan(rel, ExtensionNameIndexId, true,
+								  NULL, 1, entry);
+
+	tuple = systable_getnext(scandesc);
+
+	/* We assume that there can be at most one matching tuple */
+	if (HeapTupleIsValid(tuple))
+	{
+		result = heap_getattr(tuple, Anum_pg_extension_extversion, RelationGetDescr(rel), &is_null);
+
+		if (!is_null)
+		{
+			sql_version = strdup(TextDatumGetCString(result));
+		}
+	}
+
+	systable_endscan(scandesc);
+	heap_close(rel, AccessShareLock);
+
+	if (sql_version == NULL)
+	{
+		elog(ERROR, "Extension not found when getting version");
+	}
+	return sql_version;
+}
+
+static void inline
+do_load()
+{
+	char	   *version = extension_version();
+	char		soname[MAX_SO_NAME_LEN];
+	post_parse_analyze_hook_type old_hook;
+
+
+	snprintf(soname, MAX_SO_NAME_LEN, "%s-%s", EXTENSION_NAME, version);
+	/*
+	 * we need to capture the extension's analyzed post analyze hook, giving
```

Comment:
The logic for capturing the post-analyze-hook seems complex to me. But I couldn't figure out a better way. This is pretty well tested but perhaps we want to chicken out and say we don't support post_analyze_hooks in our extension (we can move current post_analyze_hook logic into another hook).